### PR TITLE
Fix: Component bugs

### PR DESCRIPTION
**Pull request by devanshu052000**
Closes - https://github.com/ToolJet/tj-ee/issues/5270
ee-frontend - https://github.com/ToolJet/ee-frontend/pull/625

This pull request introduces significant improvements to the marquee selection logic in the App Builder's canvas editor, ensuring that selection via drag (marquee) is correctly scoped to the intended canvas area (main, header, footer, or nested containers). It refactors the selection logic into pure, testable utility functions, adds comprehensive unit tests, and clarifies the handling of header/footer canvas IDs. Additionally, it includes minor bug fixes and improvements for button widgets.

Key changes:

**Marquee Selection Logic Refactor & Bug Fixes**
- Refactored marquee selection logic from `Selecto.jsx` into new pure utility functions (`resolveMarqueeCanvasId`, `isInMarqueeCanvas`, `mergeMarqueeSelection`) in `marqueeSelection.ts`, enabling precise scoping of selection to the correct canvas, including header/footer and nested containers. This fixes bugs where selection could leak across canvas boundaries or fail in special areas. [[1]](diffhunk://#diff-34e0fa9e7b0715316af10535fb587b36c27eec5de18fac3ffd00e4f02382c350L1-R28) [[2]](diffhunk://#diff-34e0fa9e7b0715316af10535fb587b36c27eec5de18fac3ffd00e4f02382c350L70-R61) [[3]](diffhunk://#diff-34e0fa9e7b0715316af10535fb587b36c27eec5de18fac3ffd00e4f02382c350L128-R78) [[4]](diffhunk://#diff-34e0fa9e7b0715316af10535fb587b36c27eec5de18fac3ffd00e4f02382c350L154-R103) [[5]](diffhunk://#diff-e5b6e59c84a23a2d272171593749346ff3e82af08fc1eada078c733a56e51f69R1-R54) [[6]](diffhunk://#diff-ed252970d6f9bf36e6218eb06010c0b13ec9571fbf93ec148f6568e95bb6ef48R90-R96)

- Added a comprehensive unit test suite for the marquee selection utilities, covering edge cases and previously broken behaviors such as selection in header/footer, nested containers, and shift-drag multi-selection.

**Constants Clarification**
- Introduced explicit constants `CANVAS_HEADER_ID` and `CANVAS_FOOTER_ID` in `appCanvasConstants.js` to standardize references to header/footer canvases throughout the codebase.

**Button Widget Improvements**
- Fixed the Button widget's `click` exposed variable to always reflect the current `disable` and `loading` state, preventing clicks when disabled/loading and ensuring correct event firing. [[1]](diffhunk://#diff-1def7a9307e1fe2c502a9692108c1c12ab8c08e0aac4c20e60f923001b60dad6L123-L131) [[2]](diffhunk://#diff-1def7a9307e1fe2c502a9692108c1c12ab8c08e0aac4c20e60f923001b60dad6L150-R143) [[3]](diffhunk://#diff-1def7a9307e1fe2c502a9692108c1c12ab8c08e0aac4c20e60f923001b60dad6R190-R208)
- Improved ButtonGroupV2 to update the `selected` exposed variable synchronously before firing events, ensuring that onClick actions receive the correct selection state.

**Other Minor Fixes**
- Improved label alignment in `BaseInput` for better visual consistency.
- Updated the enterprise frontend submodule pointer.

### CODE REVIEW

File: frontend/src/AppBuilder/AppCanvas/Selecto.jsx [hunk 1]
```diff
@@ -88,23 +88,22 @@ const EditorSelecto = () => {
...
+        // Scope the marquee to the canvas it was drawn on.
+        // react-selecto matches elements by bounding-rect intersection and ignores overflow clipping,
+        // so a component that is scrolled out of view inside a subcontainer can still be hit if its (clipped) rect happens to sit under the selection box.
+        // Keeping only components that belong to the starting canvas prevents the selection (and any follow-up delete) from leaking across canvas boundaries.
+        const startCanvasId = canvasStartId.current;
+        const isMarqueeOnMainCanvas = !startCanvasId || startCanvasId === 'canvas';
+        const belongsToMarqueeCanvas = (id) => {
+          const parent = getComponentDefinition(id, moduleId)?.component?.parent;
+          return isMarqueeOnMainCanvas ? !parent : parent === startCanvasId;
+        };
```
[the remark this answers is not in the document]
Comment (nakulnagargade) [reply frontend/src/AppBuilder/AppCanvas/Selecto.jsx]: @devanshu052000 Please check the co-pilot comment, other than this comment, everything else looks fine.

[the remark this answers is not in the document]
Comment (devanshu052000) [reply frontend/src/AppBuilder/AppCanvas/Selecto.jsx]: @nakulnagargade the premise doesn't hold, so no change needed for it.

The claim is that in module editor mode the root canvas id is a UUID, so startCanvasId becomes moduleId and root-level widgets get filtered out. moduleId is only ever a UUID in the viewer path, never in the editor:

- AppBuilder.jsx:30 declares Editor = ({ ..., moduleId = 'canvas', ... }) and nothing passes the prop. The only two moduleId= assignments in frontend/src + frontend/ee are GlobalSettings/index.jsx:35 ('canvas') and ee/modules/Modules/components/ModuleViewer/ModuleViewer.jsx:47 (a UUID, but that renders <Viewer>).
- Module editor mode is driven by appType === 'module' (AppLoader.jsx:21) → isModuleEditor, not by moduleId. The module editor's root canvas is still data-parentId="canvas".
- EditorSelecto only mounts under currentMode === 'edit' (AppCanvas.jsx:349), so the UUID case can't reach this code.

File: frontend/src/AppBuilder/Widgets/Button.jsx [hunk 2]
```diff
@@ -194,13 +187,25 @@ export const Button = function Button(props) {
...
+  const canClick = !disable && !loading;
+
+  useEffect(() => {
+    setExposedVariable('click', async function () {
+      if (canClick) {
+        fireEvent('onClick');
+      }
+    });
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [canClick]);
```
[the remark this answers is not in the document]
Comment (devanshu052000) [reply frontend/src/AppBuilder/Widgets/Button.jsx:199]: The window is real but narrow: the effect only re-runs when canClick actually flips, not on every render, and the realistic caller — components.button1.click() from an action — reads the exposed value from the store at call time rather than holding a reference. Leaving as-is; happy to move to a ref-backed stable function in a follow-up.

File: frontend/src/AppBuilder/AppCanvas/Selecto.jsx [hunk 3]
```diff
@@ -67,49 +41,24 @@ const EditorSelecto = () => {
...
     (e) => {
-      const realCanvasEl = e.inputEvent.target.closest('.real-canvas');
-      const canvasSelectEndId = realCanvasEl ? realCanvasEl.getAttribute('data-parentId') : null;
-      const isCanvasSelectStartEndSame = canvasStartId.current === canvasSelectEndId;
       let isMultiSelect = null;
-      let selectedIds = e.added.map((el, index) => {
+
+      const selectedIds = e.added.map((el, index) => {
         const id = el.getAttribute('widgetid');
         isMultiSelect = e.inputEvent.shiftKey || (!e.isClick && index != 0);
```
Comment (nakulnagargade) [inline frontend/src/AppBuilder/AppCanvas/Selecto.jsx:48]: This makes `isMultiSelect` true for the second and every subsequent hit of any non-click drag, even when Shift is not pressed. For example, select a header widget, then make an ordinary main-canvas drag over two widgets: `mergeMarqueeSelection` receives `true` and preserves the header selection, so the cross-canvas selection/delete risk remains. Since `scopedIds` already contains every hit in the drag, append-to-existing behavior should be driven by Shift only. Please cover this through `onSelectEnd`, not only by calling the pure merge helper.

File: frontend/src/AppBuilder/AppCanvas/Utils/__tests__/marqueeSelection.spec.js [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    test('excludes main-canvas widgets while the marquee is in the header', () => {
+      expect(isInMarqueeCanvas(undefined, 'canvas-header')).toBe(false);
+    });
+  });
+});
+
+describe('mergeMarqueeSelection', () => {
+  test('a plain drag replaces the selection', () => {
+    expect(mergeMarqueeSelection(['a', 'b'], ['old'], false)).toEqual(['a', 'b']);
```
Comment (nakulnagargade) [inline frontend/src/AppBuilder/AppCanvas/Utils/__tests__/marqueeSelection.spec.js:110]: This assertion supplies `false` directly, so it cannot catch the production wiring above deriving `isMultiSelect === true` for an ordinary drag containing two targets. Please add a component-level regression that sends a non-Shift `selectEnd` event with two added targets plus an existing selection from another canvas and verifies that the old selection is replaced.

File: frontend/src/AppBuilder/Widgets/Button.jsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (nakulnagargade) [reply frontend/src/AppBuilder/Widgets/Button.jsx:199]: The stale window is not limited to callers retaining an old function reference. The exposed setters update React state, but the stored `click` function is replaced only by this later effect. A sequential programmatic flow such as `await components.button1.setDisable(true); await components.button1.click()` can therefore read and invoke the previous `canClick === true` closure before the rerender/effect completes. That means the PR claim that `click` always reflects the current disabled/loading state is not fully met. A ref-backed guard also needs the exposed setters/property synchronization paths to update that ref synchronously.

**Comment by nakulnagargade** [comment]
Review verification summary:

- The new marquee helper suite passes locally: 21/21 tests.
- `test:layout` and `git diff --check` pass.
- The GitHub Frontend tests check has not executed the suite; it is failing because the PR lacks the `run-frontend-tests` label.
- The new tests cover only the pure marquee helpers. There is no regression coverage for the actual `EditorSelecto` event wiring, Button disabled/loading timing, ButtonGroupV2 event ordering, or BaseInput label/clear-button positioning.
- Focused lint was inconclusive locally because the installed ESLint and `@typescript-eslint` plugin versions disagree over the removed `ban-types` rule; this is not evidence of a PR lint failure.

The helper extraction is reasonable, but the marquee wiring and Button timing findings should be resolved before approval.
